cli/derive/src/lib.rs

```diff
@@ -0,0 +1,77 @@
+//! Crate with a proc macro for torii endpoint generation
+
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    parse_macro_input, Ident, LitInt, Result as SynResult,
+};
+
+/// Generates warp filters for endpoints accepting functions
+/// from 2 up to `n` arguments inclusively. Currently accepted
+/// maximum is 8. Also relies on `WarpResult` custom wrapper,
+/// and thus any module using this macro should also reexport
+/// the former.  
+#[proc_macro]
+pub fn generate_endpoints_up_to(input: TokenStream) -> TokenStream {
+    let max = parse_macro_input!(input as EndpointCount);
+    // Omitting 'e' and 'f' to allow their mnemonic use in the expansion
+    let arg_names = "abcdghij"
+        .chars()
+        .map(|c| Ident::new(format!("{c}").as_str(), Span::call_site()))
+        .collect::<Vec<_>>();
+    let arg_types = "ABCDGHIJ"
+        .chars()
+        .map(|c| Ident::new(format!("{c}").as_str(), Span::call_site()))
+        .collect::<Vec<_>>();
+    let mut endpoints = Vec::new();
+
+    for count in 2..=max.0 {
+        let fun_name = Ident::new(format!("endpoint{count}").as_str(), Span::call_site());
+        let arg_names = &arg_names[..count];
+        let arg_types = &arg_types[..count];
+
+        let expanded = quote! {
+            pub fn #fun_name < O, E, F, Fut, Fil, #( #arg_types ),* > (
+                f: F,
+                router: Fil,
+            ) -> impl Filter<Extract = (WarpResult<O, E>,), Error = Rejection> + Clone
+            where
+                Fil: Filter<Extract = ( #( #arg_types ),* ), Error = Rejection> + Clone,
+                F: Fn( #( #arg_types ),* ) -> Fut + Copy + Send + Sync + 'static,
+                Fut: std::future::Future<Output = Result<O, E>> + Send,
+                #( #arg_types: Send ),*
+                {
+                    router.and_then(move | #( #arg_names ),* | async move {
+                        Ok::<_, Infallible>(WarpResult(f( #( #arg_names ),* ).await))
+                    })
+                }
+        };
+
+        endpoints.push(expanded);
+    }
+
+    quote! {
+        #( #endpoints )*
+    }
+    .into()
+}
+
+#[derive(Debug)]
+struct EndpointCount(usize);
+
+impl Parse for EndpointCount {
+    fn parse(input: ParseStream) -> SynResult<Self> {
+        let lit: LitInt = input.parse()?;
+        let count = lit.base10_parse::<usize>()?;
+        if count > 8 || count < 2 {
+            Err(syn::Error::new_spanned(
+                lit,
+                "only endpoints with a number of arguments between two and eight are supported",
```

Comment:
```suggestion
                "only endpoints with the number of arguments between two and eight (inclusive) are supported",
```

though I'd use `from ... to` same as in the function description

Reply:
Thank you, all noted